### ISSUE TITLE
Lints (result_unit_err and missing_errors_doc) not emitted for function in "impl for" block

### ISSUE THREAD

**Issue by SylvainDe** [rust-lang/rust-clippy#9118]
### Summary

Disclaimer: I'm not quite sure yet if this is a feature but the behavior got me pretty surprised.

Implementing a from_str function returning Result<Self, ()>, I get a different set of warnings depending on whether I am in a "impl for" block:

- impl for block: not lint warnings
- impl block: lint warnings

### Lint Name

result_unit_err

### Reproducer

Context: Trying to implement a simple Point function constructible from a string I wrote:

```rust
    #[derive(Debug, PartialEq, Eq, Hash)]
    pub struct Point<T> {
        pub x: T,
        pub y: T,
    }

    use std::str::FromStr;
    impl<T: std::str::FromStr> FromStr for Point<T> {
        type Err = ();
        fn from_str(s: &str) -> Result<Self, () /*Could/should be Err*/> {
            let (x, y) = s.split_once(", ").ok_or(())?;
            Ok(Self {
                x: x.parse().map_err(|_| {})?,
                y: y.parse().map_err(|_| {})?,
            })
        }
    }

```
and got no warning.

Later, I decided that I needed a bit more flexibility and wrote:

```rust
    impl<T: std::str::FromStr> Point<T> {
        pub fn from_str_with_param(s: &str, separator: &str) -> Result<Self, ()> {
            let (x, y) = s.split_once(separator).ok_or(())?;
            Ok(Self {
                x: x.parse().map_err(|_| {})?,
                y: y.parse().map_err(|_| {})?,
            })
        }
    }
```

and I was surprised to see the following lint warnings raised:

- result_unit_err,
- missing_errors_doc

They look great to me but I do not quite get why this was not raised in the original code.

_I'm pretty new to rust (which is why I rely a lot on clippy), maybe I'm solving my problems the wrong way. I'm much interested on any comment._

### Version

```text
$ rustc -Vv
rustc 1.60.0 (7737e0b5c 2022-04-04)
binary: rustc
commit-hash: 7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c
commit-date: 2022-04-04
host: x86_64-unknown-linux-gnu
release: 1.60.0
LLVM version: 14.0.0
```

**Comment by Jarcho** [comment]
Lints about function signatures don't lint on trait impls as there's nothing the implementer can do about them. Documentation is also ignored for them as it would be basically the same as the trait's documentation.

**Comment by SylvainDe** [comment]
This makes a lot of sense. Thank you very much!
